File: src/main/java/org/lvtinger/paladin/web/filter/PaladinFilter.java
package org.lvtinger.paladin.web.filter;

import org.apache.commons.lang3.StringUtils;
import org.lvtinger.paladin.web.core.PaladinThreadLocal;

import javax.servlet.*;
import javax.servlet.annotation.WebFilter;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;

@WebFilter(filterName = "paladinFilter", urlPatterns = "/*")
public class PaladinFilter implements Filter {
    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        try {
            storeToken(request);
            chain.doFilter(request, response);
        } catch (Exception ignored) {

        } finally {
            cleanToken();
        }
    }

    private void storeToken(ServletRequest request) {
        HttpServletRequest httpServletRequest = (HttpServletRequest) request;

        String token = httpServletRequest.getHeader("token");
        if (StringUtils.isEmpty(token)) {
            return;
        }
        PaladinThreadLocal.getInstance().set(token);
    }

    private void cleanToken() {
        PaladinThreadLocal.getInstance().remove();
    }
}

File: src/main/kotlin/org/risesun/paladin/web/core/filter/PaladinFilter.kt
package org.risesun.paladin.web.core.filter

import org.risesun.paladin.web.core.PaladinHttpRequestContext
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.annotation.WebFilter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
@WebFilter(filterName = "paladinFilter", urlPatterns = ["/**"])
@Order(1)
class PaladinFilter : OncePerRequestFilter() {
    companion object {
        val context = ThreadLocal<PaladinHttpRequestContext>()
    }

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        try {
            context.set(PaladinHttpRequestContext(request, response))
            filterChain.doFilter(request, response)
        } finally {
            context.remove()
        }
    }
}